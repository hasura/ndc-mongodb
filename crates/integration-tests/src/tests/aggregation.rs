use insta::assert_yaml_snapshot;
use serde_json::json;

use crate::graphql_query;

#[tokio::test]
async fn runs_aggregation_over_top_level_fields() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query($albumId: Int!) {
                  track(order_by: { id: Asc }, where: { albumId: { _eq: $albumId } }) {
                    milliseconds
                    unitPrice
                  }
                  trackAggregate(
                    filter_input: { order_by: { id: Asc }, where: { albumId: { _eq: $albumId } } }
                  ) {
                    _count
                    milliseconds {
                      avg
                      max
                      min
                      sum
                    }
                    unitPrice {
                      _count
                      _count_distinct
                    }
                  }
                }
            "#
        )
        .variables(json!({ "albumId": 9 }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_extended_json_representing_mixture_of_numeric_types() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      avg
                      _count
                      max
                      min
                      sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "decimal|double|int|long" }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_mixture_of_numeric_and_null_values() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query ($types: String!) {
                  extendedJsonTestDataAggregate(
                    filter_input: { where: { type: { _regex: $types } } }
                  ) {
                    value {
                      avg
                      _count
                      max
                      min
                      sum
                      _count_distinct
                    }
                  }
                  extendedJsonTestData(where: { type: { _regex: $types } }) {
                    type
                    value
                  }
                }
            "#
        )
        .variables(json!({ "types": "double|null" }))
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn returns_null_when_aggregating_empty_result_set() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              moviesAggregate(filter_input: {where: {title: {_eq: "no such movie"}}}) {
                runtime {
                  avg
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn returns_zero_when_counting_empty_result_set() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              moviesAggregate(filter_input: {where: {title: {_eq: "no such movie"}}}) {
                _count
                title {
                  count
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn returns_zero_when_counting_nested_fields_in_empty_result_set() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              moviesAggregate(filter_input: {where: {title: {_eq: "no such movie"}}}) {
                awards {
                  nominations {
                    count
                    _count
                  }
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn aggregates_nested_field_values() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              moviesAggregate(
                filter_input: {where: {title: {_in: ["Within Our Gates", "The Ace of Hearts"]}}}
              ) {
                tomatoes {
                  viewer {
                    rating {
                      avg
                    }
                  }
                  critic {
                    rating {
                      avg
                    }
                  }
                }
                imdb {
                  rating {
                    avg
                  }
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}
