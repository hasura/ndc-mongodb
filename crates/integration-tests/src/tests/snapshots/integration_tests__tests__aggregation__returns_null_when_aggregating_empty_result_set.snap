---
source: crates/integration-tests/src/tests/aggregation.rs
expression: "graphql_query(r#\"\n            query {\n              moviesAggregate(filter_input: {where: {title: {_eq: \"no such movie\"}}}) {\n                runtime {\n                  avg\n                }\n              }\n            }\n            \"#).run().await?"
---
data:
  moviesAggregate:
    runtime:
      avg: ~
errors: ~
